src/operator/tensor/elemwise_unary_op_basic.cc

```diff
@@ -121,7 +121,30 @@ The storage type of ``sigmoid`` output is always dense
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseOut{"_backward_sigmoid"});
 
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU(_backward_sigmoid,
-                                               unary_bwd<mshadow_op::sigmoid_grad>);
+                                               unary_bwd<mshadow_op::sigmoid_grad>)
+.set_attr<nnvm::FGradient>("FGradient",
+    [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+      // n->inputs[0] : y_grad
+      // n->inputs[1] : f(x) = sigmoid(x)
+      // ograds[0] : head_grads
+      // f''(x) = f'(x) * (1 - 2*f(x))
+      auto ones = MakeNode("ones_like", n->attrs.name + "_grad_ones", {n->inputs[1]}, nullptr, &n);
+      const std::unordered_map<std::string, std::string> args = {{"scalar", "2.0"}};
+      auto two_y = MakeNode("_mul_scalar", n->attrs.name + "_mul_two", {n->inputs[1]}, &args, &n);
+      auto one_minus_two_y = MakeNode("elemwise_sub", n->attrs.name + "_grad_sub",
+                                    {nnvm::NodeEntry{ones}, nnvm::NodeEntry{two_y}}, nullptr, &n);
+      auto grad_grad_mid = MakeNode("elemwise_mul", n->attrs.name + "_grad_mul",
+                                    {n->inputs[0], nnvm::NodeEntry{one_minus_two_y}}, nullptr, &n);
+      // when building gradient graph, the backward node of n->inputs[1] will be
+      // added to the graph again, therefore f`(x) will be multiplied
```

Comment:
Doesn't this behaviour seem a bit odd? Is this actually the expected behaviour? What would have happened if this was `split` like function where we would have had many outputs, backward of all outputs be added? Actually I am confused by the behaviour.

Reply:
Yes, this is actually the expected behavior. The nnvm graph will perform a DFS traverse when performing backward pass. Since the input to this backward_sigmoid node is an output of another node sigmoid, during the backward pass the gradient function sigmoid will be invoked.

See:

This is to collect dependent nodes in the graph during RecordOp:
https://github.com/apache/incubator-mxnet/blob/master/src/imperative/imperative.cc#L180

This is to actually perform the backward pass:
https://github.com/dmlc/tvm/blob/21935dcbf56ad3bd66ebff9891a6bc3865b8106d/nnvm/src/pass/gradient.cc#L126

https://github.com/dmlc/tvm/blob/21935dcbf56ad3bd66ebff9891a6bc3865b8106d/nnvm/src/pass/gradient.cc#L190

Reply:
I also didn't get that. A drawing might help. I would expect that you multiply with the backward node itself "n"

Reply:
I guess that is what he meant as 
```
      // n->inputs[0] : y_grad
      // n->inputs[1] : f(x) = sigmoid(x)
      // ograds[0] : head_grads
      // f''(x) = f'(x) * (1 - 2*f(x))
```
Backward of node `n->inputs[1]` is the Node `n` itself.

One up for the visual drawing/graph of what is actually happening.

Reply:
![IMG-3494](https://user-images.githubusercontent.com/6807113/59903557-24aa9200-93b6-11e9-9762-3276865b123b.JPG)

Reply:
Here is the entire computation graph including forward pass and the second order backward pass. You can dump out the text version using 
`export MXNET_MEM_PLAN_VERBOSE_LOGGING=1`

Reply:
> Backward of node `n->inputs[1]` is the Node `n` itself.

`n` here is the backward node _backward_sigmoid.
n->inputs[1] is the forward node sigmoid.
During the second `autograd.grad()` function, it will actually trigger the FGradient function of `n->inputs[1]` which is another _backward_sigmoid node in addition to `n`. So that is why you will see two `_back_sig` nodes shown in the computation graph above.

Reply:
It will be easier to understand if you modify the FGradient function of _backward_sigmoid with only ograd[0] as input and dump out the graph. It will help you understand better.

```suggestion
      std::vector<nnvm::NodeEntry> ret;
      ret.emplace_back(ograds[0]);  // this output is not passed out if gradient w.r.t x only
      ret.emplace_back(ograds[0]);  // instead of passing back a complicated function, simply pass back the head_grad
```

Reply:
Thanks, I got it. Thanks for the explanation in person ;-)

Reply:
Thanks I'll look at it today.

Reply:
I kind of see from the graph you have drawn and dumped graph as to what is happening. I still strongly that it is the incorrect behaviour. 

To check that I modified the test code slightly to also check for the 3rd order gradient as it should come for free since the second order are composed from differentiable functions.

Have tested for `sin`, `cos`, `log`, `sigmoid`. Out of them only `sigmoid` fails for the third order.

I have also attached the hand computation of third-order for `sigmoid`, so please verify that just to make sure that it is not incorrect.

For some reason it is rotated.
![3_rd_ord](https://user-images.githubusercontent.com/19503980/59962962-271e0080-950a-11e9-91b8-cf32584bdb0a.jpeg)


```python

import math
from mxnet import nd, autograd
from mxnet.test_utils import assert_almost_equal, random_arrays, rand_shape_nd
from common import with_seed


@with_seed()
def test_sin():
    def sin(x):
        return nd.sin(x)

    def grad_grad_op(x):
        return -nd.sin(x)
    
    def grad_grad_grad_op(x):
        return -nd.cos(x)

    for dim in range(1, 5):
        shape = rand_shape_nd(dim)
        array = random_arrays(shape)
        check_second_order_unary(array, sin, grad_grad_op, grad_grad_grad_op)


@with_seed()
def test_cos():
    def cos(x):
        return nd.cos(x)

    def grad_grad_op(x):
        return -nd.cos(x)
    
    def grad_grad_grad_op(x):
        return nd.sin(x)

    for dim in range(1, 5):
        shape = rand_shape_nd(dim)
        array = random_arrays(shape)
        check_second_order_unary(array, cos, grad_grad_op, grad_grad_grad_op)

@with_seed()
def test_log():
    def log(x):
        return nd.log(x)

    def grad_grad_op(x):
        return -1/(x**2)

    def grad_grad_grad_op(x):
        return 2/(x**3)

    for dim in range(1, 5):
        shape = rand_shape_nd(dim)
        array = random_arrays(shape)
        check_second_order_unary(array, log, grad_grad_op, grad_grad_grad_op)


@with_seed()
def test_sigmoid():
    def sigmoid(x):
        return nd.sigmoid(x)

    def grad_op(x):
        return sigmoid(x) * (1 - sigmoid(x))

    def grad_grad_op(x):
        return grad_op(x) * (1 - 2 * sigmoid(x))

    def grad_grad_grad_op(x):
        return grad_grad_op(x) - 2 * ( grad_op(x)**2 + grad_grad_op(x) * sigmoid(x))

    for dim in range(1, 5):
        shape = rand_shape_nd(dim)
        array = random_arrays(shape)
        check_second_order_unary(array, sigmoid, grad_grad_op, grad_grad_grad_op)


def check_second_order_unary(x, op, grad_grad_op, grad_grad_grad_op):
    x = nd.array(x)
    grad_grad_x = grad_grad_op(x)
    grad_grad_grad_x = grad_grad_grad_op(x)
    x.attach_grad()

    # Manual head_grads.
    y_grad = nd.random.normal(shape=x.shape)
    head_grad_grads = nd.random.normal(shape=x.shape)

    # Perform compute.
    with autograd.record():
        y = op(x)
        x_grad = autograd.grad(heads=y, variables=x, head_grads=y_grad,
                               create_graph=True, retain_graph=True)[0]
        
        x_grad_grad = autograd.grad(heads=x_grad, variables=x, head_grads=head_grad_grads,
                                create_graph=True, retain_graph=True)[0]
    x_grad_grad.backward()

    # Compute expected values.
    expected_grad_grad = grad_grad_x.asnumpy() * head_grad_grads.asnumpy() * \
        y_grad.asnumpy()

    expected_grad_grad_grad = grad_grad_grad_x.asnumpy() * head_grad_grads.asnumpy() * \
        y_grad.asnumpy()

    # Validate the gradients.
    assert_almost_equal(expected_grad_grad, x_grad_grad.asnumpy())
    assert_almost_equal(expected_grad_grad_grad, x.grad.asnumpy())

if __name__ == '__main__':
    import nose
    nose.runmodule()
```

Reply:
I suspect the discrepancy in 3rd order gradient is not because an error in my implementation of grad_grad_input, but because I did not return back the first grad_grad in correctly. After all, the first output may be useful in calculating higher order even if they are not visible output at Python level. I will look into it.

Reply:
@kshitij12345 I have fixed the issue. The result can pass your test now. Please review again. Thanks!

Reply:
LGTM.